sklearn/exceptions.py

```diff
@@ -0,0 +1,144 @@
+"""
+The :mod:`sklearn.exceptions` module includes all custom warnings and error
+classes used across scikit-learn.
+"""
+
+
+__all__ = ['NotFittedError',
+           'FitFailedWarning',
+           'ChangedBehaviorWarning',
+           'DataConversionWarning',
+           'EfficiencyWarning']
+
+
+class NotFittedError(ValueError, AttributeError):
+    """Exception class to raise if estimator is used before fitting.
+
+    This class inherits from both ValueError and AttributeError to help with
+    exception handling and backward compatibility.
+
+    Examples
+    --------
+    >>> from sklearn.svm import LinearSVC
+    >>> from sklearn.exceptions import NotFittedError
+    >>> try:
+    ...     LinearSVC().predict([[1, 2], [2, 3], [3, 4]])
+    ... except NotFittedError as e:
+    ...     print(repr(e))
+    ...                        # doctest: +NORMALIZE_WHITESPACE +ELLIPSIS
+    NotFittedError('This LinearSVC instance is not fitted yet',)
+    """
+
+
+# TODO Import from model_selection
+class FitFailedWarning(RuntimeWarning):
+    """Warning class used if there is an error while fitting the estimator.
+
+    This Warning is primarily used with meta estimators like GridSearchCV,
+    RandomizedSearchCV etc. to warn when there is an error while fitting the
+    estimator.
+
+    Examples
+    --------
+    >>> from sklearn.grid_search import GridSearchCV
+    >>> from sklearn.svm import LinearSVC
+    >>> from sklearn.exceptions import FitFailedWarning
+    >>> import warnings
+    >>> warnings.simplefilter('always', FitFailedWarning)
+    >>> gs = GridSearchCV(LinearSVC(), {'C': [-1, -2]}, error_score=0)
+    >>> X, y = [[1, 2], [3, 4], [5, 6], [7, 8], [8, 9]], [0, 0, 0, 1, 1]
+    >>> with warnings.catch_warnings(record=True) as w:
+    ...     try:
+    ...         gs.fit(X, y)   # This will raise a ValueError since C is < 0
+    ...     except ValueError:
+    ...         pass
+    ...     print("The warning raised was : %r" % repr(w[-1].message))
```

Comment:
`%r` *and* `repr` seems redundant, and I expect the class name is being hidden here because of the `.message`. `'%r' % w[-1]` should suffice, I suspect.